### PR TITLE
docker/config - get all credentials

### PR DESCRIPTION
**Pull request by vrothberg**
Add means to get all credentials and fix a minor issues.

I'm currently working on the Podman v2 API and need to throw auth.jsons over the wire.

### CODE REVIEW

**Comment by vrothberg** [comment]
@mtrmac @rhatdan PTAL

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
quick rant: I really really don't like "search registries". They're inflicting headaches on a regular basis.

> I’m really not happy about this — it exposes too much about what “all” credentials look like, and how they are structured, and it makes it easier for callers to extract the data “for caching” and to do their own lookups, risking inconsistency and losing our ability to change the lookup semantics in the future.
> 
> Also consider interactive credential helpers, like GUI keyrings that may require extra passphrase inputs for high-risk keys. Just asking for all secrets when they may not be needed is not at all ideal.

We could limit support to auth files for now and call it `AuthFileToConfigMap` or similar.

> Sure, the server side must accept the projectatomic/docker map (and that does not need this API), but is there any chance the new callers would instead use a callback-based system, where the server-side search code only asks for credentials for registries one at a time, as the search code tries registries in turn? I realize that would make the API so much more complex … OTOH some infrastructure for interactive operation must already exist for `podman run -it` to work.

I don't think we can implement that in the time frame we need auth to work for v2. We had to wire in such call backs into c/image and get that through the entire stack up to libpod and then wire that into pretty much all endpoints.

> Or, maybe, provide a “get relevant search list” operation in the Podman API, so that the API client can only ask for credentials that could possibly be relevant for the operation in question. (That could still require user interactivity for all credentials on the search path, even if most of them were not necessary for that particular image.)

To double check: The server would expose an API that would give us a list of all search registries plus their mirrors. The client would then assemble the credential map based on this list and wrap it in the X-Registry-Auth header. Right?

@rhatdan PTAL.

**Comment by vrothberg** [comment]
Closing in favour of https://github.com/containers/image/pull/933. Thanks for the thoughtful comments!

File: pkg/docker/config/config.go [hunk 1]
```diff
@@ -248,6 +284,13 @@ func readJSONFile(path string, legacyFormat bool) (dockerConfigFile, error) {
 		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
 	}
 
+	if auths.AuthConfigs == nil {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+	}
+	if auths.CredHelpers == nil {
+		auths.CredHelpers = make(map[string]string)
+	}
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply pkg/docker/config/config.go:292]: Added tests in https://github.com/containers/image/pull/942. Apologies for the cross posts.

File: pkg/docker/config/config.go [hunk 2]
```diff
@@ -73,6 +73,42 @@ func SetAuthentication(sys *types.SystemContext, registry, username, password st
...
+	for registry, credHelper := range auths.CredHelpers {
+		username, password, err := getAuthFromCredHelper(credHelper, registry)
+		if err != nil {
+			return nil, err
+		}
+
+		conf := types.DockerAuthConfig{Username: username, Password: password}
+		authConfigs[registry] = conf
+	}
+
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply pkg/docker/config/config.go:108]: The kernel keyring is currently disabled. Not sure how the future will look like.

**Pull request closed without merging**